File: week-08/script1.pl
#!/usr/bin/perl

use strict;
use warnings;

# Check args
if (@ARGV < 1) {
    die "Usage: $0 file1 [file2 ...]\n";
}

foreach my $input_file (@ARGV) {
    open my $fh, '<', $input_file or die "Cannot open file $input_file: $!\n";

    my @data = <$fh>;
    close $fh;

    my $output = create_report(@data);

    my $output_file = $input_file;
    $output_file =~ s/\..+$/\.oceny/;
    open my $out_fh, '>', $output_file or die "Cannot open file $output_file for writing: $!\n";
    print $out_fh @data, $output;
    close $out_fh;

    print "Saved data from $input_file to $output_file\n";
}

sub create_report {
    my @data = @_;

    my %grades;
    my $name;

    foreach my $line (@data) {
        my @elements = split /\s+/, $line;

        my $name = ucfirst lc "$elements[0]";
        my $surname = ucfirst lc "$elements[1]";

        my $full_name = "$name $surname";
        my $grade = process_grade($elements[2]);
        push @{$grades{$full_name}}, $grade;
    }
    
    my $output = "\n";
    foreach my $person (keys %grades) {
        my $average = calculate_average(@{$grades{$person}});
        $output .= "$person Lista ocen: @{$grades{$person}} Średnia: $average\n";
    }

    return $output;
}

sub process_grade {
    my ($grade) = @_;
    my $add = 0;

    $grade =~ s/([+-])(\d)/$2$1/ if $grade =~ /[+-]\d/;

    if ($grade =~ /\+$/) { $add += 0.25; }
    if ($grade =~ /-$/)  { $add -= 0.25; }

    $grade =~ s/\+//g;
    $grade =~ s/\-//g;

    if ($grade =~ /^(\d+(\.\d+)?)$/) {
        $grade = $1;

        if ($grade >= 1 && $grade <= 6) {
            return $grade+$add;
        }
    } else {
        warn "Wrong format: $grade\n";
    }

    return;
}

sub calculate_average {
    my @grades = @_;
    my $sum = 0;
    foreach my $grade (@grades) {
        $sum += $grade;
    }
    my $average = @grades ? $sum / @grades : 0;
    return sprintf("%.2f", $average);
}

File: week-11/script1.py
#!/usr/bin/env python3

import re

def main():
    # Check args
    import sys
    if len(sys.argv) < 2:
        print("Usage:", sys.argv[0], "file1 [file2 ...]")
        sys.exit(1)

    for input_file in sys.argv[1:]:
        with open(input_file, 'r') as fh:
            data = fh.readlines()

        output = create_report(data)

        output_file = re.sub(r'\..+$', '.oceny', input_file)
        with open(output_file, 'w') as out_fh:
            out_fh.writelines(data)
            out_fh.write(output)

        print("Saved data from", input_file, "to", output_file)

def create_report(data):
    grades = {}
    name = ""

    for line in data:
        elements = line.split()
        name = f"{elements[0].capitalize()} {elements[1].capitalize()}"
        grade = process_grade(elements[2])
        grades.setdefault(name, []).append(grade)

    output = "\n"
    for person, person_grades in grades.items():
        average = calculate_average(person_grades)
        output += f"{person} Lista ocen: {person_grades} Średnia: {average:.2f}\n"

    return output

def process_grade(grade):
    add = 0

    match = re.search(r'([+-])(\d)', grade)
    if match:
        grade = match.group(2) + match.group(1) if match.group(1) in ['+', '-'] else grade

    if grade.endswith('+'):
        add += 0.25
    elif grade.endswith('-'):
        add -= 0.25

    grade = grade.replace('+', '').replace('-', '')

    try:
        grade = float(grade)
        if 1 <= grade <= 6:
            return grade + add
    except ValueError:
        print("Wrong format:", grade)

    return None

def calculate_average(grades):
    if grades:
        return sum(grades) / len(grades)
    return 0

if __name__ == "__main__":
    main()
